packages/api/src/controllers/sensorController.js

```diff
@@ -0,0 +1,208 @@
+        required: false,
+      },
+      Brand: {
+        key: 'brand',
+        parseFunction: (val) => val.trim(),
+        validator: (val) => val.length <= 100,
+        required: false,
+      },
+      Model: {
+        key: 'model',
+        parseFunction: (val) => val.trim(),
+        validator: (val) => val.length <= 100,
+        required: false,
+      },
+      Hardware_version: {
+        key: 'hardware_version',
+        parseFunction: (val) => val.trim(),
+        validator: (val) => val.length <= 100,
+        required: false,
+      },
+    });
+    if (errors.length > 0) {
+      res.status(400).send({ errors });
+    } else {
+      console.log(data);
+      res.status(200).send('Successfully uploaded!');
+    }
+  },
+
+  deleteSensor() {
+    return async (req, res) => {
+      const trx = await transaction.start(Model.knex());
+      try {
+        const isDeleted = await baseController.delete(sensorModel, '2', req, { trx });
+        await trx.commit();
+        if (isDeleted) {
+          res.sendStatus(200);
+        } else {
+          res.sendStatus(404);
+        }
+      } catch (error) {
+        //handle more exceptions
+        res.status(400).json({
+          error,
+        });
+      }
+    };
+  },
+
+  editSensor() {
+    return async (req, res) => {
+      try {
+        res.status(200).send('OK');
+      } catch (error) {
+        //handle more exceptions
+        res.status(400).json({
+          error,
+        });
+      }
+    };
+  },
+
+  getSensorsByFarmId() {
+    return async (req, res) => {
+      try {
+        // TODO: Add this back
+        // const { farm_id } = req.body.farm_id;
+        // if (!farm_id){
+        //     return res.status(400).send('No country selected');
+        // }
+        const data = await baseController.getByFieldId(sensorModel, 'farm_id', 'Testing');
+        res.status(200).send(data);
+      } catch (error) {
+        //handle exceptions
+        res.status(400).json({
+          error,
+        });
+      }
+    };
+  },
+
+  addReading() {
+    return async (req, res) => {
+      try {
+        res.status(200).send('OK');
+      } catch (error) {
+        //handle more exceptions
+        res.status(400).json({
+          error,
+        });
+      }
+    };
+  },
+
+  getAllReadingsBySensorId() {
+    return async (req, res) => {
+      try {
+        res.status(200).send('OK');
+      } catch (error) {
+        //handle more exceptions
+        res.status(400).json({
+          error,
+        });
+      }
+    };
+  },
+};
+
+/**
+ * Parses the csv string into an array of objects and an array of any lines that experienced errors.
+ * @param {String} csvString
+ * @param {Object} mapping - a mapping from csv column headers to object keys, as well as the validators for the data in the columns
+ * @param {String} delimiter
+ * @returns {Object<data: Array<Object>, errors: Array<Object>>}
+ */
+
+const parseCsvString = (csvString, mapping, delimiter = ',') => {
```

Comment:
Are we doing this validation on the backend because it may take time? I'm just interested in the rationale behind doing on backend vs the frontend.

Reply:
Oh I see, this is just parsing the data, never mind

Reply:
Wait no, I'm still confused. Are we validating the data twice?

Reply:
Yes we are doing it on both backend and frontend